dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/CategoryDimensionStore.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2004-2024, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics;
+
+import java.util.Collection;
+import java.util.List;
+import org.hisp.dhis.category.CategoryDimension;
+import org.hisp.dhis.category.CategoryOption;
+import org.hisp.dhis.common.GenericStore;
+
+/**
+ * @author david mackessy
+ */
+public interface CategoryDimensionStore extends GenericStore<CategoryDimension> {
+
+  /**
+   * Gets all {@link CategoryDimension}s that reference any of the supplied {@link CategoryOption}s
+   *
+   * @param categoryOptions to search for
+   * @return matching {@link CategoryDimension}s
+   */
+  List<CategoryDimension> getByCategoryOption(Collection<CategoryOption> categoryOptions);
```

Comment:
Still not a fan of using full objects as parameters to build restrictions. It is often unnecessary costly forcing the caller to load them by ID (how they would be specified in a API) and also leaves room for errors when using uninitialized objects as arguments.

Reply:
Yep, if this is something we want to move to as a general rule from now on, great. As long as we consistently try to enforce it. Maybe some communication to the team/wider team would help get this message across also?
So are we saying for stores and services? Or just services as you mention in another comment?

Reply:
as discussed, we'll try to use UIDs in services and UID string values in the stores. Will update.